File: go/src/medium/213.house-robber-ii.go
package medium

import "math"

func Rob(nums []int) int {
	l := len(nums)
	if l == 0 {
		return 0
	}
	if l == 1 {
		return nums[0]
	}
	return int(math.Max(float64(helper(nums, 0, l-2)), float64(helper(nums, 1, l-1))))
}

func helper(nums []int, l int, h int) int {
	prev, cur := 0, 0

	for i := l; i <= h; i++ {
		temp := int(math.Max(float64(cur), float64(prev+nums[i])))
		prev, cur = cur, temp
	}
	return cur
}

File: archive/python/old/213.house-robber-ii.py
#
# @lc app=leetcode id=213 lang=python3
#
# [213] House Robber II
#

# @lc code=start
class Solution:
    def rob(self, nums: List[int]) -> int:
        if not nums:
            return 0
        if len(nums) == 1:
            return nums[0]
        return max(self.helper(nums, 0, len(nums)-2), self.helper(nums, 1, len(nums)-1))
    def helper(self, nums: List[int], start: int, end: int) -> int:
        pre, cur = 0, 0
        for i in range(start, end):
            temp = max(pre + nums[i], cur)
            pre = cur
            cur = temp
        return cur
# @lc code=end

